### ISSUE TITLE
'The binary operator Equal is not defined...' with strongly typed keys

### ISSUE THREAD

**Issue by rick-palmsens** [dotnet/efcore#23116]
I have a strongly typed guid key:
```C#
public struct EntityReference<TEntity> : IEquatable<EntityReference<TEntity>>
    where TEntity : class, IEntity
{
    public EntityReference(Guid value)
    {
        Value = value;
    }

    public Guid Value { get; set; }
    
    public bool Equals(EntityReference<TEntity> other)
    {
        return Value.Equals(other.Value);
    }
}
```

Which I register like this:
```C#
public virtual void Configure(EntityTypeBuilder<TEntity> builder)
{
    builder
        .HasKey(entity => entity.Id);

    builder
        .Property(entity => entity.Id)
        .HasConversion(new EntityReferenceConverter<TEntity>());
}
```

With this converter:
```C#
public class EntityReferenceConverter<TEntity> : ValueConverter<EntityReference<TEntity>, Guid>
    where TEntity : class, IEntity
{
    public EntityReferenceConverter():
        base(entityReference => entityReference.Value,
             guid => new EntityReference<TEntity>(guid))
    {
    }
}
```

I set the values of the guids myself.

I get the following exception when I try to get an entity from a `DbSet`:

> System.InvalidOperationException: 'The binary operator Equal is not defined for the types 'System.Nullable`1[PalmSens.Core.Domain.Values.EntityReference`1[PalmSens.Core.Domain.Entities.Session]]' and 'System.Nullable`1[PalmSens.Core.Domain.Values.EntityReference`1[PalmSens.Core.Domain.Entities.Session]]'.'

What am I doing wrong?

### Stack Trace
```
   at System.Linq.Expressions.Expression.GetEqualityComparisonOperator(ExpressionType binaryType, String opName, Expression left, Expression right, Boolean liftToNull)
   at System.Linq.Expressions.Expression.Equal(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.ExpandingExpressionVisitor.ExpandNavigation(Expression root, EntityReference entityReference, INavigation navigation, Boolean derivedTypeConversion)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.IncludeExpandingExpressionVisitor.ExpandIncludesHelper(Expression root, EntityReference entityReference)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.IncludeExpandingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.ExpandingExpressionVisitor.Expand(Expression expression, Boolean applyIncludes)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.PendingSelectorExpandingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryTranslationPreprocessor.Process(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.FirstAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at PalmSens.Core.Application.Queries.GetSession.Handler.<Handle>d__1.MoveNext() in C:\Repositories\PalmSens\trunk\Shared\PalmSens.Core.Application\Queries\Sessions\GetSession.cs:line 68
```

### Version Info
EF Core version: 3.1.9
Database provider: Microsoft.EntityFrameworkCore.Sqlite
Target framework: .NET Standard 2.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.7.6

**Comment by rick-palmsens** [comment]
Greatly simplifying `EntityReference` until only a very minimal readonly struct was left seems to have fixed the issue. My best guess is that it had something to do with the covariant `IEntityReference` interface that got removed.

**Comment by smitpatel** [comment]
Duplicate of #19407

**Comment by rick-palmsens** [comment]
@smitpatel So am I correct to assume that it is probably fixed in EF Core 5.0?
